crates/circuit/src/circuit_data.rs

```diff
@@ -152,14 +179,80 @@ impl CircuitData {
         let args = {
             let self_ = self_.borrow();
             (
-                self_.qubits.cached().clone_ref(py),
-                self_.clbits.cached().clone_ref(py),
+                (!self_.qubits.is_empty()).then_some(self_.qubits.bits().clone()),
+                (!self_.clbits.is_empty()).then_some(self_.clbits.bits().clone()),
                 None::<()>,
                 self_.data.len(),
                 self_.global_phase.clone(),
             )
         };
-        (ty, args, None::<()>, self_.try_iter()?).into_py_any(py)
+        let state = {
+            let borrowed = self_.borrow();
+            (
+                borrowed.qregs.registers().to_vec(),
+                borrowed.cregs.registers().to_vec(),
+                borrowed.qubit_indices.cached(py).clone_ref(py),
+                borrowed.clbit_indices.cached(py).clone_ref(py),
+            )
+        };
+        (ty, args, state, self_.try_iter()?).into_py_any(py)
+    }
+
+    pub fn __setstate__(slf: &Bound<CircuitData>, state: CircuitDataState) -> PyResult<()> {
+        let mut borrowed_mut = slf.borrow_mut();
+        // Add the registers directly to the `RegisterData` struct
+        // to not modify the bit indices.
+        for qreg in state.0.into_iter() {
+            borrowed_mut.qregs.add_register(qreg, false)?;
+        }
+        for creg in state.1.into_iter() {
+            borrowed_mut.cregs.add_register(creg, false)?;
+        }
+
+        // After the registers are added, reset bit locations.
+        borrowed_mut.qubit_indices = BitLocator::from_py_dict(&state.2)?;
+        borrowed_mut.clbit_indices = BitLocator::from_py_dict(&state.3)?;
+
+        Ok(())
+    }
+
+    /// Returns the a list of registered :class:`.QuantumRegisters` instances.
```

Comment:
I would reword this to

```suggestion
    /// The a list of registered :class:`.QuantumRegisters` instances.
```

since it's a getter so it'll act like an attribute not a method.

Reply:
Since in 0f8cf6593.